content-src/lib/selectLayoutRender.js

```diff
@@ -1,79 +1,70 @@
-import {createSelector} from "reselect";
+export const selectLayoutRender = (state, rickRollCache) => {
+  const {layout, feeds, spocs} = state;
+  let spocIndex = 0;
 
-export const selectLayoutRender = createSelector(
-  // Selects layout, feeds, spocs so that we only recompute if
-  // any of these values change.
-  [
-    state => state.DiscoveryStream.layout,
-    state => state.DiscoveryStream.feeds,
-    state => state.DiscoveryStream.spocs,
-  ],
-
-  // Adds data to each component from feeds. This function only re-runs if one of the inputs change.
-  // TODO: calculate spocs
-  function layoutRender(layout, feeds, spocs) {
-    let spocIndex = 0;
-
-    function maybeInjectSpocs(data, spocsConfig) {
-      if (data &&
-          spocsConfig && spocsConfig.positions && spocsConfig.positions.length &&
-          spocs.data.spocs && spocs.data.spocs.length) {
-        const recommendations = [...data.recommendations];
-        for (let position of spocsConfig.positions) {
-          let rickRoll = Math.random();
-          if (spocs.data.spocs[spocIndex] && rickRoll <= spocsConfig.probability) {
-            recommendations.splice(position.index, 0, spocs.data.spocs[spocIndex++]);
-          }
+  function maybeInjectSpocs(data, spocsConfig) {
+    if (data &&
+        spocsConfig && spocsConfig.positions && spocsConfig.positions.length &&
+        spocs.data.spocs && spocs.data.spocs.length) {
+      const recommendations = [...data.recommendations];
+      for (let position of spocsConfig.positions) {
+        // Cache random number for a position
+        if (!rickRollCache[position.index]) {
```

Comment:
Is position.index unique? (I forget)

Reply:
I think a new component could in theory have a new set of positions with the same index as before.

Reply:
yes for a specific component picked from layout , it should be unique.
```
{
          "type": "CardGrid",
          "properties": {
            "items": 3
          },
          "header": {
            "title": ""
          },
          "feed": {
            "embed_reference": null,
            "url": "https://getpocket.cdn.mozilla.net/v3/firefox/global-recs?..."
          },
          "spocs": {
            "probability": 1,
            "positions": [
              {
                "index": 2
              }
            ]
          }
        },


```

Reply:
> I think a new component could in theory have a new set of positions with the same index as before.
yes, in case of  a layout with multiple components say comp 1 and comp 2., comp 2 can have spoc position with same index as comp 1 and will be using the cached random value .

Reply:
This is probably not ideal though?

Reply:
My understanding spoc item should hide and display based off probability passed in the feed. 
For layout with multiple components, spoc items in different components displayed at the same position will show/hide based off spocs config probability. 

The only caveat is if spoc items in different components has same config probability and same config position , they will always show and hide at the same time (which can happen in random cases even with existing implementation in master). 

For  Hero, List and Card Grid with different number of items, whats the chance for spoc items showing up at same position and with same probability. May be  we should run it by Product and get their take on it. Is it worth supporting this use case by saving rick rolls by component and spoc positions.

Reply:
I don't think I follow what you mean by same config probability?

Reply:
> I don't think I follow what you mean by same config probability?

It's spocs -> probability received in layout for a respective component https://github.com/mozilla/activity-stream/pull/4858#discussion_r267532491